torch/_dynamo/guards.py

```diff
@@ -611,6 +642,145 @@ def _produce_guard_code(
         )
 
 
+# Common Sub-Expression Elimination for Python expressions.
+#
+# There are 2 steps to this pass:
+#     1. Count the frequency of each sub-expression (i.e. inner
+#        node in the AST tree)
+#
+#     2. Replace those that occur more than once by a fresh variable 'v'.
+#        'v' will be defined in the 'preface' list (output argument to
+#        'NodeTransformer')
+class PyExprCSEPass:
+    IGNORED_NODE_TYPES = (
+        # Proxy nodes
+        # i.e. nodes that result in the same unparsed string as their children
+        ast.Expression,
+        ast.Index,
+        ast.Expr,
+        ast.Module,
+        # Leaf Expression nodes
+        ast.Name,
+        ast.Constant,
+        # Expr-Context nodes
+        ast.Load,
+        ast.Store,
+        ast.Del,
+        # Bool Operation nodes
+        ast.And,
+        ast.Or,
+        # Arithmetic Operation nodes
+        ast.Add,
+        ast.Sub,
+        ast.Mult,
+        ast.MatMult,
+        ast.Div,
+        ast.Mod,
+        ast.Pow,
+        ast.LShift,
+        ast.RShift,
+        ast.BitOr,
+        ast.BitXor,
+        ast.BitAnd,
+        ast.FloorDiv,
+        # Unary Operation nodes
+        ast.Invert,
+        ast.Not,
+        ast.UAdd,
+        ast.USub,
+        # Compare Operation nodes
+        ast.Eq,
+        ast.NotEq,
+        ast.Lt,
+        ast.LtE,
+        ast.Gt,
+        ast.GtE,
+        ast.Is,
+        ast.IsNot,
+        ast.In,
+        ast.NotIn,
+    )
+
+    class CSEVisitor(ast.NodeVisitor):
+        IGNORE = object()
+
+        def __init__(self) -> None:
+            self._expr_counter: Dict[str, int] = collections.defaultdict(lambda: 0)
+
+        def visit(self, node: ast.AST) -> Any:
+            if _ast_unparse_implemented(node) and not isinstance(
+                node, PyExprCSEPass.IGNORED_NODE_TYPES
+            ):
+                self._expr_counter[_ast_unparse(node)] += 1
+            super().visit(node)
+
+    class CSETransformer(ast.NodeTransformer):
+        def __init__(
+            self,
+            expr_counter: Dict[str, int],
+            preface: List[str],
+            gen_name: Callable[[], str],
+        ) -> None:
+            super().__init__()
+            self._expr_counter = expr_counter
+            self._preface = preface
+            self._gen_name = gen_name
+            self._expr_to_name: Dict[str, str] = {}
+
+        def visit(self, node: ast.AST) -> Any:
+            if _ast_unparse_implemented(node) and not isinstance(
+                node, PyExprCSEPass.IGNORED_NODE_TYPES
+            ):
+                expr = _ast_unparse(node)
```

Comment:
NB: because visit recursively traverses children, this is quadratic in expression size